### DISCUSSION TITLE
ui.stat_list_item clickable action

### DISCUSSION THREAD

**Discussion by CH3V4L13R** [h2oai/wave#728] [Q&A] [answered]
I'm trying to figure out how to add a specific action done when a row of the ui.stat_list_item is clicked. 
my table is like this:
```py
ui.stat_list_item(
                label=nb_unshipped_list['SALES_ID'][i],
                name=f'order_clicked_{str(i)}',
                caption=get_order_list_items(nb_unshipped_list['SALES_ITEMS'][i]),
                value=nb_unshipped_list['SALES_DATE'][i],
                aux_value=nb_unshipped_list['EC'][i],
                value_color=get_order_waiting_color(nb_unshipped_list['SALES_DATE'][i])) for i in range(len(nb_unshipped_list))
        ],
```
I want the user to be able to have information displayed when a specific row is clicked but the q.args only contain the name of the list row. 
Is it possible to get others info like the label, caption info of the specific clicked row inside the q.args as well so those data can be processed later?

My main issue being how to check for which row is clicked since the q.args.order_clicked_0, q.args.order_clicked_1, etc... name for the  number of row is dynamic. And doing something like this doesn't work:
```py
list_clicked_order = "list_clicked_" + str(1)
if q.args.list_clicked_order:
```
....

The name of the row being fixed doesn't make sense as all rows would have the same action behavior.
Thanks, any feedback or help are welcome...

**Comment by mturoci** [comment]
Hi @CH3V4L13R, good question. The quick and ugly workaround could be something like this:
```py
from h2o_wave import main, app, Q, ui


# Your custom data structure.
class DataObj:
    def __init__(self, prop1, prop2, prop3):
        self.prop1 = prop1
        self.prop2 = prop2
        self.prop3 = prop3


# Fill the dictionary - key has to be same as name of the stat_list_item.
obj_dict = {}
for i in range(10):
    obj_dict[i] = DataObj(i, i, i)


@app('/demo')
async def serve(q: Q):
    q.page['list'] = ui.stat_list_card(box='1 1 5 5', title='Title', items=[
        ui.stat_list_item(name=f'{idx}', label=str(obj.prop1)) for idx, obj in obj_dict.items()
    ])

    # Ugly workaround for getting q.args.check
    for key in obj_dict.keys():
        if str(key) in q.args:
            print(f'Row {key} was clicked!')
            print(f'Here is the associated data: {obj_dict[key]}')

    # Dummy update - ignore.
    q.page['dummy'].items = []
    await q.page.save()
```

@lo5 maybe we could add same pattern matching for `args`  as we have for routing, wdyt?

**Reply by CH3V4L13R** [reply]
@mturoci Thank you for taking the time to send me a workaround. It's definitively less ugly than my personal code ;)
I will take some time this weekend to upgrade my code with your solution and see how it goes. Thanks again!

**Reply by lo5** [reply]
@CH3V4L13R @mturoci The correct solution (1-liner) is to provide a `name` to the card, e.g. `ui.stat_list_card(..., name='foo', ...)`, so that `q.args.foo` is set to the name of the selected item.

**Comment by mtanco** [comment]
I wanted to mention that because `q.args` is an expando it can behave like both an object and a dictionary, so this:

```python
list_clicked_order = "list_clicked_" + str(1)
if q.args.list_clicked_order:
```

can be handled with:

```py
list_clicked_order = "list_clicked_" + str(1)
if q.args[list_clicked_order] is not None:
```

In this case @mturoci 's solution is the way to go, checking if each `key is in q.args` but you can also do for the same effect:
```py
if q.args[ str(key)] is not None:
            print(f'Row {key} was clicked!')
            print(f'Here is the associated data: {obj_dict[key]}')
```

**Comment by lo5** [comment]
The correct solution (1-liner) is to provide a `name` to the card, e.g. `ui.stat_list_card(..., name='foo', ...)`, so that `q.args.foo` is set to the name of the selected item.

**Comment by CH3V4L13R** [comment]
@mturoci @lo5 @mtanco Thank you for all your feedback. The code provided by @mturoci works perfectly. The issue is that the list contain a certain number of orders and the list is dynamically generated from that number:

`for i in range(len(nb_unshipped_list)):
    obj_dict[i] = DataObj(i)`

From that clicked row, I just need to retrieve one specific value (the order id) and can get all the other info from it later.

@mtanco I wish I won't have to check all the key in q.args via the for loop, but I don't understand well your explanation. Since I do not know how many keys will be generated or which key (list row) will be clicked, I guess doing the for loop to check is the only way to go. Please let me know if preventing the for loop seems possible.

To summarize, the only things that I needed was the class object with one variable to check which row is being clicked. 

class DataObj:
    def __init__(self, prop1):
        self.prop1 = prop1

and 

obj_dict = {}
for i in range(len(nb_unshipped_list)):
    obj_dict[i] = DataObj(i)

Thanks!

**Comment by lo5** [comment, marked as answer]
@CH3V4L13R - the solution from @mturoci is not the right approach. 

@mturoci - no workarounds/hacks are necessary. 

You do not have to loop through `q.args`.
Just provide a `name`, e.g. `ui.stat_list_card(, ... name='foo', ...)` and you can simply check the value of `q.args.foo` to know which item was selected.

**Reply by CH3V4L13R** [reply]
@lo5 the name needs to be dynamic, foo_1, foo_2... for each line generated for the stat_list_card so that each line refers to a specific info (in my case the order id). How do you retrieve the specific line number (order id) info clicked with a static name?

**Reply by CH3V4L13R** [reply]
Here is a screenshot of my stat_list_card, sorry I should have show this from the beginning.
The stat_list_card display a list of orders. When the user click on one of the line (the order), all the order info appears into a different card.

![Screenshot from 2021-07-01 09-16-17](https://user-images.githubusercontent.com/67501864/124046509-1aafd500-da4d-11eb-8e04-00802d025271.png)

**Reply by lo5** [reply]
Sure. Assuming your order IDs are `foo_1`, `foo_2`, etc., set up your card like this:

```py
my_card = ui.stat_list_card(
  box='...',
  title='...',
  name='order_id',
  items=[
    ui.stat_list_item(name='foo_1', ...),
    ui.stat_list_item(name='foo_2', ...),
    ui.stat_list_item(name='foo_3', ...),
  ]
)
```

When an item in the card is clicked, read the order ID like this:

```py
selected_order_id = q.args.order_id
```

You can then use this ID to look up order details and display it in your other card.

Hope this helps, unless I'm missing something.

**Reply by lo5** [reply]
Note that this pattern is universally followed across Wave. For example, https://github.com/h2oai/wave/blob/master/py/examples/table_select.py

**Reply by lo5** [reply]
More complete/useful example: https://wave.h2o.ai/docs/examples/issue-tracker - same pattern

**Reply by CH3V4L13R** [reply]
@lo5 thank you for the links and example, it helps me understand the pattern better. I was able to apply the right code to my use case and it's working perfectly fine now without the for loop. 
Again, @lo5 @mturoci @mtanco thank you for all the support!
